packages/sdk/src/conversation/sso/ssoExecutionDialog.ts

```diff
@@ -0,0 +1,307 @@
+   * Called when the component is ending.
+   *
+   * @param context Context for the current turn of conversation.
+   */
+  protected async onEndDialog(context: TurnContext) {
+    const conversationId = context.activity.conversation.id;
+    const currentDedupKeys = this.dedupStorageKeys.filter((key) => key.indexOf(conversationId) > 0);
+    await this.dedupStorage.delete(currentDedupKeys);
+    this.dedupStorageKeys = this.dedupStorageKeys.filter((key) => key.indexOf(conversationId) < 0);
+  }
+
+  /**
+   * If a user is signed into multiple Teams clients, the Bot might receive a "signin/tokenExchange" from each client.
+   * Each token exchange request for a specific user login will have an identical activity.value.Id.
+   * Only one of these token exchange requests should be processed by the bot. For a distributed bot in production,
+   * this requires a distributed storage to ensure only one token exchange is processed.
+   * @param context Context for the current turn of conversation.
+   * @returns boolean value indicate whether the message should be removed
+   */
+  private async shouldDedup(context: TurnContext): Promise<boolean> {
+    const storeItem = {
+      eTag: context.activity.value.id,
+    };
+
+    const key = this.getStorageKey(context);
+    const storeItems = { [key]: storeItem };
+
+    try {
+      await this.dedupStorage.write(storeItems);
+      this.dedupStorageKeys.push(key);
+    } catch (err: unknown) {
+      if (err instanceof Error && err.message.indexOf("eTag conflict")) {
+        return true;
+      }
+      throw err;
+    }
+    return false;
+  }
+
+  private getStorageKey(context: TurnContext): string {
+    if (!context || !context.activity || !context.activity.conversation) {
+      throw new Error("Invalid context, can not get storage key!");
+    }
+    const activity = context.activity;
+    const channelId = activity.channelId;
+    const conversationId = activity.conversation.id;
+    if (activity.type !== ActivityTypes.Invoke || activity.name !== tokenExchangeOperationName) {
+      throw new Error("TokenExchangeState can only be used with Invokes of signin/tokenExchange.");
+    }
+    const value = activity.value;
+    if (!value || !value.id) {
+      throw new Error("Invalid signin/tokenExchange. Missing activity.value.id.");
+    }
+    return `${channelId}/${conversationId}/${value.id}`;
+  }
+
+  private matchPattern(pattern: string | RegExp, text: string): boolean | RegExpMatchArray {
+    if (text) {
+      if (typeof pattern === "string") {
+        const regExp = new RegExp(pattern as string, "i");
+        return regExp.test(text);
+      }
+
+      if (pattern instanceof RegExp) {
+        const matches = text.match(pattern as RegExp);
+        return matches ?? false;
+      }
+    }
+
+    return false;
+  }
+
+  private shouldTrigger(patterns: TriggerPatterns, text: string): RegExpMatchArray | boolean {
+    const expressions = Array.isArray(patterns) ? patterns : [patterns];
```

Comment:
Our code use both `Error` and `ErrorWithCode`, should we only use one of them?

Reply:
OK